File: backend/src/main/java/com/company/aiinterview/chat/entity/ChatMessage.java
package com.company.aiinterview.chat.entity;

import com.company.aiinterview.interview.entity.InterviewSession;
import jakarta.persistence.*;
import lombok.*;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDateTime;

@Entity
@Table(name = "chat_messages")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@EntityListeners(AuditingEntityListener.class)
public class ChatMessage {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "session_id", nullable = false)
    private InterviewSession session;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, columnDefinition = "VARCHAR(20)")
    private Sender sender;

    @Column(name = "message_type")
    private String messageType;

    @Column(name = "turn_no")
    private Integer turnNo;

    @Column(name = "message_content", columnDefinition = "TEXT", nullable = false)
    private String messageContent;

    @Column(name = "model_name")
    private String modelName;

    @Column(name = "prompt_tokens")
    private Integer promptTokens;

    @Column(name = "completion_tokens")
    private Integer completionTokens;

    @Column(name = "response_time_ms")
    private Long responseTimeMs;

    @CreatedDate
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "suggested_answer", columnDefinition = "TEXT")
    private String suggestedAnswer;

    @Column(name = "score_clarity")
    private Integer scoreClarity;

    @Column(name = "score_technical")
    private Integer scoreTechnical;

    @Column(name = "score_confidence")
    private Integer scoreConfidence;

    @Column(name = "category_topic", columnDefinition = "VARCHAR(100)")
    private String categoryTopic;
}

File: backend-node/src/models/ChatMessage.ts
import mongoose, { Document, Schema } from 'mongoose';

export enum Sender {
  USER = 'USER',
  AI = 'AI',
  SYSTEM = 'SYSTEM'
}

export interface IChatMessage extends Document {
  session: mongoose.Types.ObjectId;
  sender: Sender;
  messageType?: string;
  turnNo?: number;
  messageContent: string;
  modelName?: string;
  promptTokens?: number;
  completionTokens?: number;
  responseTimeMs?: number;
  suggestedAnswer?: string;
  scoreClarity?: number;
  scoreTechnical?: number;
  scoreConfidence?: number;
  scoreSituation?: number;
  scoreTask?: number;
  scoreAction?: number;
  scoreResult?: number;
  categoryTopic?: string;
  createdAt: Date;
}

const chatMessageSchema: Schema = new Schema(
  {
    session: { type: Schema.Types.ObjectId, ref: 'InterviewSession', required: true },
    sender: { type: String, enum: Object.values(Sender), required: true },
    messageType: { type: String },
    turnNo: { type: Number },
    messageContent: { type: String, required: true },
    modelName: { type: String },
    promptTokens: { type: Number },
    completionTokens: { type: Number },
    responseTimeMs: { type: Number },
    suggestedAnswer: { type: String },
    scoreClarity: { type: Number },
    scoreTechnical: { type: Number },
    scoreConfidence: { type: Number },
    scoreSituation: { type: Number },
    scoreTask: { type: Number },
    scoreAction: { type: Number },
    scoreResult: { type: Number },
    categoryTopic: { type: String },
  },
  {
    timestamps: { createdAt: true, updatedAt: false }, // Only need createdAt for messages
  }
);

export const ChatMessage = mongoose.model<IChatMessage>('ChatMessage', chatMessageSchema);
